### PR TITLE
add blocking thread customize

### PR DESCRIPTION
**Pull request by fakeshadow**
## PR Type

Feature

## PR Checklist
Check your PR fulfills the following:

- [x] Documentation comments have been added / updated.
- [x] A changelog entry has been made for the appropriate packages.
- [x] Format code with the latest stable rustfmt

## Overview

Make blocking thread pool count changable at per worker level.

### CODE REVIEW

File: actix-server/src/builder.rs [hunk 1]
```diff
@@ -78,6 +78,25 @@ impl ServerBuilder {
...
+    /// # Examples:
+    /// ```rust
+    /// # use actix_server::ServerBuilder;
+    /// let builder = ServerBuilder::new()
+    ///     .workers(4) // server has 4 worker thread.
+    ///     .worker_max_blocking_threads(4); // every worker has 4 max blocking threads.
+    /// ```
+    ///
+    /// See [tokio::runtime::Builder::max_blocking_threads](tokio::runtime::Builder::max_blocking_threads)
+    /// for behavior reference.
```
Comment (robjtede) [inline actix-server/src/builder.rs:94]:
```suggestion
    /// See [tokio::runtime::Builder::max_blocking_threads] for behavior reference.
```

File: actix-server/src/builder.rs [hunk 2]
```diff
@@ -78,6 +78,25 @@ impl ServerBuilder {
         self
     }
 
+    /// Set number of max blocking threads for running blocking task on thread pool.
+    ///
+    /// * This thread pool is per worker and `NOT` shared across workers.
+    ///
+    /// # Examples:
+    /// ```rust
```
Comment (robjtede) [inline actix-server/src/builder.rs:86]:
```suggestion
    /// ```
```

File: actix-server/src/builder.rs [hunk 3]
```diff
@@ -78,6 +78,25 @@ impl ServerBuilder {
         self
     }
 
+    /// Set number of max blocking threads for running blocking task on thread pool.
```
Comment (robjtede) [inline actix-server/src/builder.rs:81]:
```suggestion
    /// Set max number of threads for each worker's blocking task thread pool.
```

File: actix-server/src/builder.rs [hunk 4]
```diff
@@ -78,6 +78,25 @@ impl ServerBuilder {
...
+    ///
+    /// * This thread pool is per worker and `NOT` shared across workers.
```
Comment (robjtede) [inline actix-server/src/builder.rs:83]:
```suggestion
    /// One thread pool is set up **per worker**; not shared across workers.
```

File: actix-server/src/worker.rs [hunk 5]
```diff
@@ -159,26 +159,61 @@ enum WorkerServiceStatus {
     Stopped,
 }
 
+// config for worker behavior passed from server builder.
```
Comment (robjtede) [inline actix-server/src/worker.rs:162]:
```suggestion
/// Config for worker behavior passed down from server builder.
```

File: actix-server/src/worker.rs [hunk 6]
```diff
@@ -159,26 +159,61 @@ enum WorkerServiceStatus {
...
+    shutdown_timeout: Duration,
+    max_blocking_threads: usize,
+}
+
+impl Default for ServerWorkerConfig {
+    fn default() -> Self {
+        Self {
+            shutdown_timeout: Duration::from_secs(30),
+            // default max_blocking_threads is the same as tokio's default
+            max_blocking_threads: 512,
```
Comment (robjtede) [inline actix-server/src/worker.rs:174]: I think it should be a bit lower to account for common case of multi-worker so as to not potentially blow up thread count. 512 / 4 sounds good. So 128.

**Review by robjtede — CHANGES_REQUESTED** [review]
looks good, just some doc nits

File: actix-server/src/worker.rs (hunk 6 above)
Comment (fakeshadow) [reply actix-server/src/worker.rs:174]: What about divide 512 with the core count num_cpus returns.

Comment (robjtede) [reply actix-server/src/worker.rs:174]: good idea, i guess we have that as a dep anyway

**Review by robjtede — APPROVED** [review]

**Comment by robjtede** [comment]
Somewhat arguable it should be `512 / #workers` but this is fine.

**Pull request merged**